Fixes #4167, Fixes #3542 Person imports via Grouper/EDS integration

tadean opened the pull request:
## Description
This draft PR uses EDS credentials stored in the `az_eds` module in #4373 and uses functionality from that PR.

A new migrate source plugin, `az_ldap` is provided that allows using a saved ldap query as a migrate source.

## Related issues
Closes #4167
Closes #3542 
#4372 (depends upon)

## How to test

- Perform testing instructions in #4373 to inject EDS credentials into your build
- Log in to the site
- Enable `az_person_eds_import`
- Visit `/admin/config/people/ldap/server` on the site
- Verify that the server labeled `EDS Service` reads `Server available`
- Visit `/admin/config/people/ldap/query`
- Click `Add LDAP Queries`
- Create an EDS Query for a group of people
  - Add a descriptive Label
  - Select `EDS Service` for `LDAP server used for query`
  - For `Base DNs to search in query` include the `Base DN for queries` from stache entry `UITS AZ Quickstart EDS Account`
  - For `Filter`, include a valid LDAP query. A simple one is `(isMemberOf=GROUPER_GROUP_NAME)`
    - replace `GROUPER_GROUP_NAME` with a desired grouper group
  - For `Attributes`, include: 
```
uid,preferredPriorityGivenName,preferredPrioritySn,employeeTitle,pronouns,employeePhone,mail
```
  - Save your query
  - At `/admin/config/people/ldap/query`, test that your query returns your desired results via the `Test` Operation
  - At `/admin/content/eds/import` select your saved Query, and for `/admin/content/eds/import` select `Import all persons`
  - Click `import`
  - Verify that your people were imported correctly
 
## Testing Moderation Drafts
  - Choose one person and make an edit to them
  - Visit `/admin/config/workflow/workflows/manage/editorial/type/node`
  - Set Person to be affected by the editorial workflow
  - Save the settings
  - Visit `/admin/content/eds/import`
  - Import your query again, once again with `Import all persons`
  - Visit the person you edited in the previous step
  - Verify that your edits are still the live revision
  - Verify that the Latest Revision is an unpublished revision in moderation state draft that reverts your changes
 
## Testing Change Tracking
  - Visit `/admin/content/eds/import`
  - Import your query again, this time with `Choose how profiles are imported` set to `Import new/changed persons`
  - Verify all people in your query are again imported
  - Import your query again, with `Choose how profiles are imported` set to `Import new/changed persons`
  - Verify that **NO** people are imported (because they are not changed)
 
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [x] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

tadean commented:
@joeparsons I temporarily reverted your suggestion about `Drupal\migrate\Attribute\MigrateSource` because it does not exist until Drupal 11 (I think). I'll add it back in once the Drupal 11 PR is merged.

tadean commented:
- Added option to allow for usage of `track_changes`
- Added functionality to allow for person migrations to participate in content moderation status so that `track_changes` imports can generate drafts with the changed information.

danahertzberg commented:
Could use list of persons with new revision (in draft or published) after import.

tadean commented:
- [ ] Add event subscriber to show imported persons
- [ ] Add optional query selector to profiles import form to serve as profiles import netid input, IF the EDS module is enabled

joeparsons commented on modules/custom/az_person/az_person_profiles_import/src/Form/AZPersonProfilesImportForm.php:
```diff
@@ -107,6 +151,30 @@ public function submitForm(array &$form, FormStateInterface $form_state): void {
+        }
+      }
+    }
+    catch (\Exception $e) {
+    }
```
```suggestion
```

joeparsons commented on modules/custom/az_person/az_person_profiles_import/src/Form/AZPersonProfilesImportForm.php:
```diff
@@ -107,6 +151,30 @@ public function submitForm(array &$form, FormStateInterface $form_state): void {
       $urls[] = $netid;
     }
 
+    try {
+      if (!empty($query)) {
+        // This service may not exist if ldap_query is not enabled.
+        // This shouldn't occur, because the element only appears if enabled.
+        $ldapQueryController = \Drupal::service('ldap.query');
+        // Attempt to execute the specified query.
+        $ldapQueryController->load($query);
+        $ldapQueryController->execute();
+        $results = $ldapQueryController->getRawResults();
+        // For each person we find, add them as a net id to the profile import.
+        foreach ($results as $result) {
+          $row = $result->getAttributes();
+          $uid = $row['uid'] ?? [];
+          $uid = reset($uid);
+          if (!empty($uid)) {
+            \Drupal::logger('my_module')->notice($uid);
+            $urls[] = trim($uid);
+          }
+        }
+      }
+    }
```
```suggestion
    if (!empty($query) && $this->ldapQueryController instanceof \Drupal\ldap_query\Controller\QueryController) {
      // Attempt to execute the specified query.
      $this->ldapQueryController->load($query);
      $this->ldapQueryController->execute();
      $results = $this->ldapQueryController->getRawResults();
      // For each person we find, add them as a net id to the profile import.
      foreach ($results as $result) {
        $row = $result->getAttributes();
        $uid = $row['uid'] ?? [];
        $uid = reset($uid);
        if (!empty($uid)) {
          \Drupal::logger('my_module')->notice($uid);
          $urls[] = trim($uid);
        }
      }
    }
```

joeparsons reviewed (commented):
Suggestion(s) for using dependency injection for optional `ldap.query` service.

tadean commented on modules/custom/az_person/az_person_profiles_import/src/Form/AZPersonProfilesImportForm.php:
```diff
@@ -107,6 +151,30 @@ public function submitForm(array &$form, FormStateInterface $form_state): void {
+          $uid = $row['uid'] ?? [];
+          $uid = reset($uid);
+          if (!empty($uid)) {
+            \Drupal::logger('my_module')->notice($uid);
```
Oops :face_exhaling: - this log statement was actually unintentionally committed. (Though, maybe we want a logger channel for some other reasons).

tadean commented:
- [ ] Change to DN-based migration id seems to have broken overlapping imports of profiles and eds imports
- [ ] Apply same phone formatting used in profiles import

tadean commented:
- Refactored fetcher custom behavior for profiles imports to shared process plugin.
- Updated with changes from main.

tadean commented:
Updated format to `PhoneNumberFormat::NATIONAL` to keep parity with current format.

tadean commented:
As follow ups from #4373, we determined the following should happen in this PR:

- [ ] EDS queries should selectively add additional automatic filters that prevent certain groups from being included (for example, excluding inactive employees)
  - [ ]  in some cases, some of these filters may be configurable as Quickstart settings
- [ ] EDS imports should respect grouper groups and student selections to indicate that certain data should suppressed at the person's choice. We intend to do this by making the source plugin aware of these attributes and selectively filter the data
- [ ] Spoke with infrastructure team to get more information and we were told that grouper group names and the list of EDS attributes are not secret, and so consequently we **can include** them in our code.

tadean commented:
Added `az_eds.config_override` service tagged with `config.factory.override` to provide notion of overriding default query settings for EDS queries.

tadean commented:
Added optout process plugin to strike certain attributes from the record if other related data-privacy attributes are encountered.

tadean commented on modules/custom/az_eds/az_eds.services.yml:
```diff
@@ -0,0 +1,6 @@
+services:
+  az_eds.config_override:
+    class: \Drupal\az_eds\AZEDSConfigOverrides
+    arguments: ['@config.factory', '@cache.data']
```
```suggestion
    arguments: ['@config.factory']
```

tadean commented:
- Added settings form
- Changed opt-out plugin to be called `az_eds_privacy`
- Added optional affiliation requirements to privacy rules
- Added filter to filter out students, configured by configuration form

danahertzberg reviewed (approved):
All worked locally for me!

tadean commented:
Removed the `AZ_EDS_ADDRESS` secret because the infrastructure team has confirmed it is public information and does not need to be implemented as a secret.

tadean commented:
Added clause that prevents the import of any student from occurring if they have `PRIMARY NAME` info release code.

danahertzberg reviewed (approved):
Works locally!

tadean commented:
Added AJAX tooltip to query config form when EDS server is selected:
<img width="1001" height="467" alt="image" src="https://github.com/user-attachments/assets/7ab00137-ce85-4da6-9602-3201ec92d86b" />

tadean commented:
- Refactored to use `BASE_FORM_ID` to avoid a generic `form_alter`
- Updated tooltip language.
- Clustered Base DN and attributes elements together
> Should the notice have a link to where the defaults are set in case someone wants to look them up?

@danahertzberg where should this link to?

tadean commented:
@camikazegreen suggested that link to documentation can be a followup after documentation for the feature becomes available.
